File: app/static/js/filter.js
async function filterData(){
    const class_id = document.getElementById("class_id").value;
    const subject_id = document.getElementById("subject_id").value;
    const topic_id = document.getElementById("topic_id").value;
    const lesson_id = document.getElementById("lesson_id").value;
    const from_date = document.getElementById("from_date").value;
    const to_date = document.getElementById("to_date").value;

    const res = await fetch(
        `/filter/?class_id=${class_id}&subject_id=${subject_id}&topic_id=${topic_id}&lesson_id=${lesson_id}&from_date=${from_date}&to_date=${to_date}`
    );

    const data = await res.json();
    renderResult(data);
}

File: app/routers/filter.py
from fastapi import APIRouter
from app.database.postgres import get_conn

router = APIRouter(prefix="/filter")

@router.get("/")
def filter_data(
    class_id: str = None,
    subject_id: str = None,
    topic_id: str = None,
    lesson_id: str = None,
    from_date: str = None,
    to_date: str = None
):
    conn = get_conn()
    cur = conn.cursor()

    query = """
        SELECT lesson_id, title, created_at
        FROM lesson
        WHERE is_deleted = false
    """
    params = []

    if class_id:
        query += " AND class_id=%s"
        params.append(class_id)

    if subject_id:
        query += " AND subject_id=%s"
        params.append(subject_id)

    if topic_id:
        query += " AND topic_id=%s"
        params.append(topic_id)

    if lesson_id:
        query += " AND lesson_id=%s"
        params.append(lesson_id)

    if from_date:
        query += " AND created_at >= %s"
        params.append(from_date)

    if to_date:
        query += " AND created_at <= %s"
        params.append(to_date)

    query += " LIMIT 100"

    cur.execute(query, tuple(params))
    rows = cur.fetchall()
    conn.close()

    return [
        {"lesson_id": r[0], "title": r[1], "created_at": str(r[2])}
        for r in rows
    ]
